### PR TITLE
Support L2-less devices with fast redirect

### LINKED ISSUE #15075: datapath: Support L2-less with fast redirect
Once #14858 has been merged, make it possible to run the fast redirect (aka BPF host routing) with L2-less devices.

From the first glance it might be trivial to add the support. In the redirect functions we should bypass setting L2 addrs if egress iface doesn't have L2 addr.

UPDATE: the kernel fix by @joamaki has been merged. What's left is detecting whether the fix is present in the kernel (maybe via a helper as we do for other features), and then enabling the functionality by removing https://github.com/cilium/cilium/blob/v1.13.0-rc5/daemon/cmd/kube_proxy_replacement.go#L455.

### PR DESCRIPTION
**Pull request by jschwinger233**
This PR adds L2 header for skb on L2-less devices before redirecting to internal veth, and enables bpf-based host routing work with wireguard as well as L2-less interfaces.

Fixes: #15075 

```release-note
Support L2-less devices with fast forward (bpf-based host routing) 
```

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

### CODE REVIEW

File: bpf/lib/l3.h [hunk 1]
```diff
@@ -14,6 +14,36 @@
...
+{
+    __u16 proto = ctx_get_protocol(ctx);
+    *l2_hdr_required = false;
+
+#ifdef IS_L3_DEV
+	if (IS_L3_DEV(ifindex)) {
+		/* If the dest dev is also an L3 dev, we skip the L2 header settings.
+		 */
+        return 0;
+    }
```
Comment (jschwinger233) [inline bpf/lib/l3.h]: indent

File: test/bpf_tests/bpf_test.go [hunk 2]
```diff
@@ -85,6 +85,9 @@ func TestBPF(t *testing.T) {
 		if !strings.HasSuffix(entry.Name(), ".o") {
 			continue
 		}
+		if entry.Name() != "tc_nodeport_l3_dev.o" {
```
Comment (jschwinger233) [inline test/bpf_tests/bpf_test.go]: remember to delete

File: bpf/bpf_host.c [hunk 3]
```diff
@@ -441,6 +441,8 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 	struct endpoint_info *ep;
 	void *data, *data_end;
 	struct iphdr *ip4;
+	bool l2_hdr_required = true;
+	int l3_off = ETH_HLEN;
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: We can reduce the scope here, and just define these variables inside the relevant section:

```
ep = lookup_ip4_endpoint(ip4);
if (ep) {
	bool l2_hdr_required = true;
	int l3_off = ETH_HLEN;

	...
}
```

File: bpf/bpf_host.c [hunk 4]
```diff
@@ -534,6 +536,19 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
...
+			return ret;
+		if (l2_hdr_required && ETH_HLEN == 0) {
+			/* l2 header is added */
+			l3_off += __ETH_HLEN;
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip4, sizeof(*ip4),
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Should we only include this section when `#ifdef ENABLE_HOST_ROUTING`?

I believe we also need the same handling in the IPv6 path.

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Thank you @jschwinger233 ! Some comments below.

File: bpf/bpf_host.c (hunk 4 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: I think currently ENABLE_HOST_ROUTING can't go along with ipv6, can it? 
host routing requires bpf masquerading: https://github.com/cilium/cilium/blob/master/daemon/cmd/kube_proxy_replacement.go#L461
bpf masquerading isn't supported for ipv6: https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L1413

Comment (julianwiedmann) [reply bpf/bpf_host.c]: Ah of course - thanks! (it's actually in the works at the moment: https://github.com/cilium/cilium/pull/23165)

So either we add the IPv6 datapath code already today (knowing that it's unused). Or we add a check/TODO at the right location in the agent code, to make sure we don't forget doing so when support for IPv6 BPF-Masq is added.

File: bpf/tests/tc_nodeport_l3_dev.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	struct tcphdr *l4;
+	void *data;
+
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+
+	/* We are building an L3 skb which doesn't have L2 header, so in theory
+	 * we need to set ctx->protocol = bpf_ntohs(ETH_P_IP), but bpf verifier
+	 * doesn't allow us to do so, and kernel also doesn't handle an L3 skb
+	 * properly (see https://elixir.bootlin.com/linux/latest/source/net/bpf/test_run.c#L1156).
```
Comment (dylandreimerink) [inline bpf/tests/tc_nodeport_l3_dev.c]: Nit: Switch the link for a permalink version.
```suggestion
	 * properly (see https://elixir.bootlin.com/linux/v6.2.1/source/net/bpf/test_run.c#L1156).
```

**Review by dylandreimerink — APPROVED** [review]
Very nice!

File: bpf/bpf_host.c (hunk 3 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: Compiler will complain "error: mixing declarations and code is a C99 extension [-Werror,-Wdeclaration-after-statement]" - -

**Comment by jschwinger233** [comment]
@rgo3 Hi Gögge, may I have some of your input here?

Let me give you some background. In this PR we enable bpf host routing to work with wireguard, and a key dependency is the helper `bpf_skb_change_head` from kernel. 

Now We want to make sure the kernel has included a specific [patch](https://lore.kernel.org/bpf/20210519154743.2554771-2-joamaki@gmail.com/) for this helper, and my current solution is to simply check if kernel version >= 5.13.0, please see the file diffs: https://github.com/cilium/cilium/pull/23935/files#diff-c1e209552c5426ba2f2e01f043275558eb9f8ea65553210d2f9808de17b59d1e

Right now we don't need to work out a perfect solution to exam the backporting situations for older kernels (maybe later we should), so I was wondering if you have some idea about the best way to do it?

File: daemon/cmd/kube_proxy_replacement.go [hunk 6]
```diff
@@ -462,15 +470,19 @@ func finishKubeProxyReplacementInit() error {
 		// All cases below still need to be implemented ...
 		case option.Config.EnableEndpointRoutes:
 			msg = fmt.Sprintf("BPF host routing is currently not supported with %s.", option.EnableEndpointRoutes)
-		case !mac.HaveMACAddrs(option.Config.GetDevices()):
-			msg = "BPF host routing is currently not supported with devices without L2 addr."
-		case option.Config.EnableWireguard:
-			msg = fmt.Sprintf("BPF host routing is currently not compatible with Wireguard (--%s).", option.EnableWireguard)
+		//
+		case !mac.HaveMACAddrs(option.Config.GetDevices()) || option.Config.EnableWireguard:
+			if kernelVersion, err := version.GetKernelVersion(); err == nil {
```
Comment (nebril) [inline daemon/cmd/kube_proxy_replacement.go]: We don't handle the `err` returned here if it's not nil. Is there any way to log it?

**Comment by rgo3** [comment]
> Now We want to make sure the kernel has included a specific [patch](https://lore.kernel.org/bpf/20210519154743.2554771-2-joamaki@gmail.com/) for this helper, and my current solution is to simply check if kernel version >= 5.13.0, please see the file diffs: https://github.com/cilium/cilium/pull/23935/files#diff-c1e209552c5426ba2f2e01f043275558eb9f8ea65553210d2f9808de17b59d1e

I think the only risk with checking for the kernel version directly is that certain things can be backported on RHEL, especially with regards to helpers, but I am not sure if that applies to patches to helpers as well. I think that is why it is generally avoided except in `pkg/datapath/linux/requirements.go`. I don't really have a better workaround though.

@borkmann do we have other instances where datapath requires a patched helper and just feature probing the helper could yield false positive results?

File: daemon/cmd/kube_proxy_replacement.go [hunk 7]
```diff
@@ -34,9 +34,17 @@ import (
...
 	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/cilium/cilium/pkg/version"
+	"github.com/cilium/cilium/pkg/versioncheck"
 	wgTypes "github.com/cilium/cilium/pkg/wireguard/types"
 )
 
+var (
+	// Host routing on the L3 devices counts on a patch for bpf_skb_change_head: https://lore.kernel.org/bpf/20210519154743.2554771-2-joamaki@gmail.com/
+	// The patch is available as of kernel 5.13-rc4, so we simply check kernel version to guarantee that.
+	hasSkbChangeHeadPatch = versioncheck.MustCompile(">=5.13.0")
```
Comment (pchaigno) [inline daemon/cmd/kube_proxy_replacement.go]: We don't usually rely on specific kernel versions because that breaks too easily (e.g., RHEL kernels). Instead we probe the specific kernel feature we need if possible, or something that was introduced at the same time (same patchset ideally, same release otherwise).

In this case, it's not a feature but a bugfix. I'm expecting it will be backported to all relevant LTS kernels (?). Do we even need to enable this conditionally? What happens if the kernel doesn't support it? How obvious is the failure today?

File: pkg/datapath/linux/config/config.go [hunk 8]
```diff
@@ -205,6 +205,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	cDefinesMap["LB4_BACKEND_MAP"] = "cilium_lb4_backends_v3"
 	cDefinesMap["LB4_REVERSE_NAT_SK_MAP"] = lbmap.SockRevNat4MapName
 	cDefinesMap["LB4_REVERSE_NAT_SK_MAP_SIZE"] = fmt.Sprintf("%d", lbmap.MaxSockRevNat4MapEntries)
+	cDefinesMap["IS_L3_DEV(ifindex)"] = "false"
```
Comment (pchaigno) [inline pkg/datapath/linux/config/config.go]: Why is this needed? That macro is already defined lower in the same function.

File: bpf/node_config.h [hunk 9]
```diff
@@ -35,6 +35,10 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
  #define SECCTX_FROM_IPCACHE fetch_u32(SECCTX_FROM_IPCACHE)
 #endif
 
+#ifndef IS_L3_DEV
+#define IS_L3_DEV(ifindex) false
+#endif
```
Comment (pchaigno) [inline bpf/node_config.h]: This is already defined on line 213 of this same file.

**Review by pchaigno — COMMENTED** [review]
Nice work on your first datapath PR! :tada: 

I think the kernel version check needs to be changed to a kernel probe or even just removed. Besides that, just nit comments.

**Comment by jschwinger233** [comment]
This PR is subject to #24180

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (jschwinger233) [reply daemon/cmd/kube_proxy_replacement.go]: > I'm expecting it will be backported to all relevant LTS kernels

This is correct, I just checked the latest LTS kernels, 4.14.307, 4.19.275,  5.4.234, 5.10.172, and all of them have included the patch.

> What happens if the kernel doesn't support it?

According to the patch commit message, without it, redirecting a packet from a L3 device such as wireguard to the veth peer device will cause skb->data to point to the middle of the IP header on entry to tcp_v4_rcv().

File: bpf/node_config.h (hunk 9 above)
Comment (jschwinger233) [reply bpf/node_config.h]: In the first place there wasn't such definition until I rebased the upstream. Let me delete the duplicate macro.

**Comment by julianwiedmann** [comment]
Note that this currently won't be covered in `ci-datapath` due to https://github.com/cilium/cilium/issues/23283.

File: bpf/bpf_host.c [hunk 10]
```diff
@@ -256,9 +256,24 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		/* Let through packets to the node-ip so they are
 		 * processed by the local ip stack.
 		 */
+		bool l2_hdr_required __maybe_unused = true;
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Please move this declaration to the top of the block, so that the comment stays with the `ENDPOINT_F_HOST` handling.

File: bpf/bpf_host.c [hunk 11]
```diff
@@ -256,9 +256,24 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
...
+		ret = maybe_add_l2_hdr(ctx, ep->ifindex, &l2_hdr_required);
+		if (ret != 0)
+			return ret;
+		if (l2_hdr_required && ETH_HLEN == 0) {
+			/* l2 header is added */
+			l3_off += __ETH_HLEN;
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip6, sizeof(*ip6),
+						      false, l3_off))
+				return DROP_INVALID;
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Don't think we need this revalidation for IPv6?

File: bpf/bpf_host.c [hunk 12]
```diff
@@ -529,9 +544,27 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 		/* Let through packets to the node-ip so they are processed by
 		 * the local ip stack.
 		 */
+		bool l2_hdr_required __maybe_unused = true;
+		int l3_off __maybe_unused = ETH_HLEN;
+
```
Comment (julianwiedmann) [inline bpf/bpf_host.c:555]: Same, please move in front of the comment.

File: bpf/lib/nodeport.h [hunk 13]
```diff
@@ -26,6 +26,7 @@
 #include "stubs.h"
 #include "proxy_hairpin.h"
 #include "fib.h"
+#include "neigh.h"
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: This is probably just leftover from pre-rebase, and can go away?

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Just some smaller ~cosmetic things to pick up during the needed rebase, thanks!

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (pchaigno) [reply daemon/cmd/kube_proxy_replacement.go]: So Cilium would likely drop those packets because they are malformed. Maybe it's fine to just enable this by default without any probing? We can ask in the community meeting if unsure.

File: bpf/bpf_host.c (hunk 11 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: Correct, since `ip6` is no more used afterwards.

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (borkmann) [reply daemon/cmd/kube_proxy_replacement.go]: > So Cilium would likely drop those packets because they are malformed. Maybe it's fine to just enable this by default without any probing? We can ask in the community meeting if unsure.

I think if we document this with regards to stable kernel requirements, this would be fine too if pkts would be dropped for the case where they are not present. At least it would be a strong enough signal.

Comment (borkmann) [reply daemon/cmd/kube_proxy_replacement.go]: The other option is to probe for a helper which went into the same release as Jussi's fixes and only enable it then. This would ignore the stable releases for the time being, but at least it would ensure there are no gaps in between.

File: bpf/bpf_host.c (hunk 11 above)
Comment (borkmann) [reply bpf/bpf_host.c]: And for the way out (L2 -> L3), this is being taken care of via maybe_add_l2_hdr in fib_redirect right?

File: bpf/bpf_host.c [hunk 14]
```diff
@@ -538,12 +554,30 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
...
+			return ret;
+		if (l2_hdr_required && ETH_HLEN == 0) {
+			/* l2 header is added */
+			l3_off += __ETH_HLEN;
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip4, sizeof(*ip4),
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+#endif
```
Comment (borkmann) [inline bpf/bpf_host.c:575]: Could this check be done unconditionally instead of via ifdefs?

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (pchaigno) [reply daemon/cmd/kube_proxy_replacement.go]: I'm proposing dropping the probe mostly because it's fairly complicated to implement (for a probe) and this feels like a fairly corner case scenario already.

Comment (jschwinger233) [reply daemon/cmd/kube_proxy_replacement.go]: Sure. To implement a probe looks fun, I could do it in future if really necessary.

File: bpf/bpf_host.c (hunk 14 above)
Comment (jschwinger233) [reply bpf/bpf_host.c:575]: I think it could be done, since there's no side effect for normal L2-L2 redirect, or even host routing is off, this `maybe_add_l2_hdr` still does no harm. Julian, how do you like it? @julianwiedmann

Comment (julianwiedmann) [reply bpf/bpf_host.c:575]: I don't see the advantage - if we only require it for host-routing, why burden the BPF program on eg. older kernels with the additional complexity?

File: bpf/bpf_host.c (hunk 11 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: I think for the wireguard scenario where the typical return path is like lxc(L2) -> cilium_wg0(L3),  we don't have to take care of the extra L2 header in ctx, the helper `ctx_redirect(ctx, WG_IFINDEX, 0)` can redirect an L2 packet to L3 device, is that right?

**Comment by pchaigno** [comment]
/test-1.24-5.4

**Comment by jschwinger233** [comment]
> Restarting because [it failed](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.24-kernel-5.4/1163/) with what looks like two different new flakes. Rerunning to confirm. If it passes, we need to open issues for those flakes.
> 
> /test-1.24-5.4

Now it results with even more failures....

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (jschwinger233) [reply daemon/cmd/kube_proxy_replacement.go]: I'm going to resolve this conversation by now. The conclusion so far is to make no check and to presume costumers are using TLS kernels which have been backported with necessary patches. If it turns out to be needed in future, I can implement a feature probe then.

File: bpf/node_config.h (hunk 9 above)
Comment (jschwinger233) [reply bpf/node_config.h]: The duplicated macro has been removed and I'll resolve this.

File: bpf/bpf_host.c (hunk 14 above)
Comment (jschwinger233) [reply bpf/bpf_host.c:575]: For the benefit of less code complexity, I decide to keep these `ifdef`. If any other thought coming into mind, please unresolve the conversation.

File: pkg/datapath/linux/config/config.go (hunk 8 above)
Comment (jschwinger233) [reply pkg/datapath/linux/config/config.go]: Removed, thanks for reminding.

**Comment by jschwinger233** [comment]
> Restarting because [it failed](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.24-kernel-5.4/1163/) with what looks like two different new flakes. Rerunning to confirm. If it passes, we need to open issues for those flakes.
> 
> /test-1.24-5.4

For this failure, I can make it pass locally with 2 changes:

1. cherrypick #24343 to avoid operator deadlock
2. add `time.Sleep(10*time.Second) in the [AfterEach](https://github.com/cilium/cilium/blob/v1.14.0-snapshot.0/test/k8s/datapath_configuration.go#L513) to await policy to be essentially deleted from bpf map.

So I presume the e2e test failures I hit before were both flakes, although the policy one is maybe not recorded before.

I can fix these "delete CRDs without waiting" race condition in e2e test in a separate PR later.

cc @julianwiedmann

File: daemon/cmd/kube_proxy_replacement.go (hunk 7 above)
Comment (julianwiedmann) [reply daemon/cmd/kube_proxy_replacement.go]: Just for documentation purposes:

- `bpf: Set mac_len in bpf_skb_change_head` landed in kernel `5.13`
- bpf-host-routing is available since kernel 5.10
- between 5.10 and 5.13, the only kernel still maintained on kernel.org is `5.10`.
- `bpf: Set mac_len in bpf_skb_change_head` was backported to `5.10.42`

**Comment by julianwiedmann** [comment]
>     2. add `time.Sleep(10*time.Second)` in the [AfterEach](https://github.com/cilium/cilium/blob/v1.14.0-snapshot.0/test/k8s/datapath_configuration.go#L513) to await policy to be truly deleted from bpf map.
> 

:tada: would be great to have a follow-on issue for this aspect!

File: bpf/bpf_host.c [hunk 15]
```diff
@@ -538,12 +550,30 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
...
+
 		return ipv4_local_delivery(ctx, ETH_HLEN, secctx, ip4, ep,
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]:
```suggestion
		return ipv4_local_delivery(ctx, l3_off, secctx, ip4, ep,
```

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Just one tiny fix-up needed. CI is all green already :heavy_check_mark:

File: bpf/bpf_host.c (hunk 15 above)
Comment (jschwinger233) [reply bpf/bpf_host.c]: Of course! Thanks!

**Review by julianwiedmann — APPROVED** [review]
Looks great, thank you @jschwinger233 ! :rocket:

File: bpf/tests/tc_nodeport_l3_dev.c [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+#define TEST_IP_LOCAL		v4_pod_one
+#define TEST_IP_REMOTE		v4_pod_two
+#define TEST_IPV6_LOCAL		v6_pod_one
+#define TEST_IPV6_REMOTE	v6_pod_two
+#define TEST_LXC_ID_LOCAL	233
+
+/* We wanted to tail call handle_policy from bpf_lxc, but at present it's
+ * impossible to #include both bpf_host.c and bpf_lxc.c at the same time.
+ * Therefore, we created a stud, mock_hanle_policy, to simply check if the
+ * our skb reaches there.
```
Comment (qmonnet) [inline bpf/tests/tc_nodeport_l3_dev.c:26]: Not worth a respin, but if you need to update your PR for another reason:

```suggestion
 * Therefore, we created a stud, mock_hanle_policy, to simply check if
 * our skb reaches there.
```

There's another typo in the description of your first commit: `lagacy`.

**Review by qmonnet — APPROVED** [review]
Nice work!

**Comment by julianwiedmann** [comment]
The failing update test is known flake https://github.com/cilium/cilium/issues/24379: 

> /home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.19/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:515
> migrate-svc restart count values do not match
> Expected
>     <int>: 0
> to be identical to
>     <int>: 13
> /home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.19/src/github.com/cilium/cilium/test/k8s/updates.go:346

**Comment by jibi** [comment]
Marking this for v1.13 backporting as another PR relies on some these changes.
Will only backport 46ea80c3264af64f947d6584f1500288a2310ee8
